logr.go

```diff
@@ -248,10 +279,40 @@ func (l Logger) WithSink(sink LogSink) Logger {
 // Glass" in the package documentation). Normally the sink should be used only
 // indirectly.
 type Logger struct {
-	sink            LogSink
-	level           int
-	ctx             context.Context
-	contextHandlers *[]ContextHandler // Has to be a pointer to keep the struct comparable.
+	sink  LogSink
+	level int
+
+	// This is a pointer to keep Logger as small as possible.
+	// The effect of having to allocate contextKV instances
+	// gets mitigated via sync.Pool.
+	contextKV *contextKV
+}
+
+// contextKV is shared by different Logger instances. It combines everything
+// that is needed to retrieve key/value pairs from a context.
+type contextKV struct {
+	ctx      context.Context
+	handlers []ContextHandler
+}
+
+var contextKVPool = contextKVPoolType{
+	Pool: sync.Pool{
+		New: func() interface{} {
+			return new(contextKV)
+		},
+	},
+}
+
+type contextKVPoolType struct {
+	sync.Pool
+}
+
+func (pool *contextKVPoolType) Get(ctx context.Context, handlers []ContextHandler) *contextKV {
```

Comment:
Having an item in a sync.Pool does not prevent its being garbage collected. So if the contextKV instance in the pool gets garbage-collected, the context and sink also will be (potentially...). Because of that I hadn't bother with clearing the struct on Put. There may be value in dropping pointers to them on Put because that makes these instance available for garbage-collection sooner. I'll try whether it makes a noticeable difference.

I'll also try without it. The amount of allocations will go up, so I expect the difference to be notiecable.

Reply:
Clearing on Put made no difference. I'm using it anyway because it feels cleaner.

Using the pool is faster than not using it:

    NewContext100PercentDisabled/simple-36                     72.2µs ± 1%    66.8µs ± 3%   -7.48%  (p=0.008 n=5+5)
    NewContext100PercentDisabled/simple-36                     32.0kB ± 0%    27.2kB ± 0%  -14.88%  (p=0.008 n=5+5)
    NewContext100PercentDisabled/simple-36                        500 ± 0%       400 ± 0%  -20.00%  (p=0.008 n=5+5)

It's not a huge difference, but IMHO worthwhile.

Reply:
After taking a more thorough look at this change I noticed that `contextKV` instances were getting cleared on `Put` that were still in use by other Logger instances. This implies that `Put` got called too early, I just hadn't noticed.

With hindsight, that makes sense: when one `Logger` instance stops using a certain `contextKV` instance, it has no knowledge of other `Logger` instances.

To avoid the extra alloc during FromContext (the subjectively most frequent call that currently triggers a recreation of `contextKV`), at least the ctx field would have to be in Logger again. When adding caching, `With*` calls also need a new instance. So perhaps storing everything in Logger is the right approach after all - I'll have to do a final comparison at the end.

Reply:
yeah, I was trying to put my finger on what didn't feel right still, and this might be it.  It feels like thebest we can do is 1 alloc per `WithContext()` (assuming we use a small fixed array for small sets of handlers)